uast/transformer/semantic.go

```diff
@@ -194,8 +195,10 @@ func (c *commentElems) Split(text string) bool {
 
 	c.Suffix = ""
 	if i >= 0 {
-		c.Suffix = text[i+1:]
-		text = text[:i+1]
+		// compute the unicode length (could be more than just one byte)
+		_, ulen := utf8.DecodeRune([]byte(text[i:]))
```

Comment:
This works exactly for 1 Unicode rune at the end of the string. You'll need to loop here to decode all runes.

Since the same bug is probably there for prefixes as well, I think it may be better to convert string to runes first (`[]rune(s)`) and then apply other logic to runes instead of bytes.

Reply:
I am going to answer the second part first. I have thought about that, and with prefixes there are no problems (I think), because in the case of `Prefix`es:

```
	i := strings.IndexFunc(text, func(r rune) bool {
		return !c.isTab(r)
	})

	c.Prefix = ""

	if i >= 0 {
		c.Prefix = text[:i]
		text = text[i:]
	}
```

Note that `i` is actually the position (in bytes of the first byte of the rune), so for prefixes it would work fine, because we take until that first byte and then the rest of the string.

To the first observation you made: `i` is the first byte position for the last thing that is not a tab. The problem taking suffixes is that I need to include that character into the Suffix, so taking from the next byte onwards would just simply not work when the unicode character contains more than one byte. I have changed the test to include more utf8 (with more than one byte) characters at the end of the string. It works :smile:

Reply:
Also, yes, it would better to apply the logic to `[]rune` instead of `[] byte`. But we do not have methods to work with `[] rune` (prefixes, suffixes, trim and such) and a lot to work with strings. Also if we convert to `[]rune` we would have to do it at the beginning and stick with `[] rune` to avoid a lot of conversions back and forth from `strings`. For now this fix works well, and I would not want to touch anything else in the method really.